syntax = "proto3";

package stat_service;

option go_package = "./proto";

service StatService {
    rpc GetStatPost(GetStatPostRequest) returns (StatPostResponse);
    rpc GetTopPosts(GetTopPostsRequest) returns (GetTopPostsResponse);
    rpc GetTopNUsers(GetTopNUsersRequest) returns (GetTopNUsersResponse);
}

message GetStatPostRequest {
    int64 Post_id = 1;
}

message GetTopPostsRequest {
    bool Views = 1;
}

message GetTopNUsersRequest {
    int64 Top_N = 1;
}

message StatPostResponse {
    int64 Post_id = 1;
    int64 Likes = 2;
    int64 Views = 3;
}

message GetTopPostsResponse {
    repeated AbstractStatPost Posts = 1;
}

message GetTopNUsersResponse {
    repeated TopUsersLikes Users = 1;
}

message Post {
    int64 Id = 1;
    string Title = 2;
    int64 User_id = 3;
    string Text_description = 4;
    string Post_time = 5;
}

message TopUsersLikes {
    int64 User_id = 1;
    int64 Total_likes = 2;
}

message AbstractStatPost {
    int64 Amount = 1;
    int64 Id = 2;
    int64 Author = 3;
}